crates/runtime/src/instance/allocator/pooling/uffd.rs

```diff
@@ -0,0 +1,833 @@
+
+            return Some(AddressLocation::TablePage {
+                page_addr: (table_start + (page_index * self.page_size)) as _,
+                len: self.page_size,
+            });
+        }
+
+        // Check for a stack location
+        if addr >= self.stacks_start && addr < self.stacks_end {
+            let index = (addr - self.stacks_start) / self.stack_size;
+            let stack_start = self.stacks_start + (index * self.stack_size);
+            let stack_offset = addr - stack_start;
+            let page_offset = (stack_offset / self.page_size) * self.page_size;
+
+            return Some(AddressLocation::StackPage {
+                page_addr: (stack_start + page_offset) as _,
+                len: self.page_size,
+                index,
+                guard_page: stack_offset < self.page_size,
+            });
+        }
+
+        None
+    }
+}
+
+unsafe fn wake_guard_page_access(
+    uffd: &Uffd,
+    page_addr: *const u8,
+    len: usize,
+) -> Result<(), String> {
+    // Set the page to NONE to induce a SIGSEV for the access on the next retry
+    region::protect(page_addr, len, region::Protection::NONE)
+        .map_err(|e| format!("failed to change guard page protection: {}", e))?;
+
+    uffd.wake(page_addr as _, len).map_err(|e| {
+        format!(
+            "failed to wake page at {:p} with length {}: {}",
+            page_addr, len, e
+        )
+    })?;
+
+    Ok(())
+}
+
+unsafe fn initialize_wasm_page(
+    uffd: &Uffd,
+    instance: &Instance,
+    page_addr: *const u8,
+    memory_index: usize,
+    page_index: usize,
+) -> Result<(), String> {
+    if let Some(MemoryInitialization::Paged { page_size, map }) =
+        &instance.module.memory_initialization
+    {
+        let memory_index = DefinedMemoryIndex::new(memory_index);
+        let memory = instance.memory(memory_index);
+        let pages = &map[memory_index];
+        debug_assert_eq!(WASM_PAGE_SIZE % page_size, 0);
+
+        let count = WASM_PAGE_SIZE / page_size;
+        let start = page_index * count;
+
+        for i in start..start + count {
+            let dst = memory.base.add(i * page_size);
+
+            match pages.get(i) {
+                Some(Some(data)) => {
+                    log::trace!(
+                        "copying page initialization data from {:p} to {:p} with length {}",
+                        data,
+                        dst,
+                        page_size
+                    );
+
+                    // Copy the page data without waking
+                    uffd.copy(data.as_ptr() as _, dst as _, *page_size, false)
+                        .map_err(|e| {
+                            format!(
+                                "failed to copy page from {:p} to {:p} with length {}: {}",
+                                data, dst, page_size, e
+                            )
+                        })?;
+                }
+                _ => {
```

Comment:
Could this have a comment that `Some(None)` means that we just didn't have a data initializer for this page in memory but we did have an initializer for a later page, and then `None` means that we didn't have an initializer for this page or anything afterwards.

Reply:
I'll add that.